### DISCUSSION TITLE
Python version of FLAME GPU 2

### DISCUSSION THREAD

**Discussion by daniele-baccega** [FLAMEGPU/FLAMEGPU2#1034] [Q&A] [answered]
Hi,
what do you mean with **_transpiled_** in "agent functions expressed as Python syntax which are transpiled to C++"?

**Comment by Robadob** [comment, marked as answer]
Transpiling is the process of taking source code written in one language and transforming it into another language that has a similar level of abstraction.

In this case FLAMEGPU2's Python transpiler takes an agent function expressed in a pythonic syntax e.g.

```python
@pyflamegpu.agent_function
def add_func(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageNone):
    x = pyflamegpu.getVariableInt("x")
    pyflamegpu.setVariableInt("x", x + 20)
    return pyflamegpu.ALIVE
```

You pass it to the transpiler

```python
        func_translated = pyflamegpu.codegen.translate(add_func)
```

This will store a string with the C++ equivalent agent function inside `func_translated`

e.g.
```c++
FLAMEGPU_AGENT_FUNCTION(add_func, flamegpu::MessageNone, flamegpu::MessageNone) {
    int x = FLAMEGPU->getVariable<int>("x");
    FLAMEGPU->setVariable<int>(x + 20);
    return flamegpu::ALIVE;
}
```

Which can then be passed to FLAMEGPU as a runtime compiled (RTC) function for use in a model. Or you could just look at the generated string, if you want to understand how the C++ equivalent would be structured.

```py
func = a.newRTCFunction("add_func", func_translated)
```

There are some limitations to this, e.g. you can't start using `numpy` functions or similar python packages, although we support several `numpy` types as they map better to the full range of C/C++ data types available.

Hope that helps.

*In drafting this reply, I spotted a mistake in one of the related code examples in the documentation, [I will get that fixed ASAP](https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/135).*
